content/source/docs/enterprise/sentinel/import/index.html.md

```diff
@@ -24,100 +24,178 @@ see [Mocking Terraform Sentinel Data](../mock.html).
+                                # If not, print violation message
+                                if r.applied.instance_type not in allowed_types {
+                                        print("EC2 instance", address, "has attribute",
+                                                r.applied.instance_type, "that is not in the list", allowed_types)
+                                        validated = false
+                                }
+                        }
+                }
+        }
+
+        return validated
+}
+```
+
+This function calls the `find_resources_from_plan` and `get_instance_address` functions described above.
+
+The boolean variable `validated` is initially set to `true`, but it is set to `false` if any resource instance violates the condition requiring that the `instance_type` attribute be in the `allowed_types` list. Since the function returns `true` or `false`, it can be called inside Sentinel rules.
+
+Note that this function prints a warning message for each resource instance that violates the condition. This allows the writer of a Sentinel rule that calls it to tell a user about **all** resource instances that violate it. This is much better than just reporting the first violation since a user with multiple violations learns about all of them in a single shot.
+
+### Call Functions From Rules
+
+Finally, having re-used the standardized `find_resources_from_plan` and `get_instance_address` functions and having written your own function to validate that resources instances of a specific type satisfy some condition, you can define a list with allowed values and write a rule that calls your function:
 
 ```python
 # Allowed Types
 allowed_types = [
-	"t2.small",
-	"t2.medium",
-	"t2.large",
+        "t2.small",
+        "t2.medium",
+        "t2.large",
 ]
 
+# Call the validation function and assign results
+instance_types_validated = validate_instance_types(allowed_types)
+
 # Rule to restrict instance types
 instance_type_allowed = rule {
-	all get_resources("aws_instance") as r {
-		r.applied.instance_type in allowed_types
-	}
+        instance_types_validated is true
+}
+
+# Main rule
+main = rule {
+        instance_type_allowed is true
+}
+
+```
+
+Note that you could leave out the `instance_type_allowed` rule and place `instance_types_validated is true` directly inside your `main` rule. You could also leave out `is true` in both rules; we included it here to be more explicit.
+
+### Validating Multiple Conditions in a Single Policy
+
+If you want a policy to validate multiple conditions against resources of a specific type, we recommend that you use a single function to iterate over all resource instances of that type and make this function return a list of boolean values, using one for each condition.  You can then use multiple Sentinel rules that evaluate those boolean values or evaluate them all in your `main` rule. Here is a partial example:
```

Comment:
I understand _why_ this is necessary in some instances - namely to reduce evaluation time on larger plans by removing the need to scan over all of the data multiple times. I wish it didn't need to be done this way as it circumvents some of Sentinel's tracing functionality.

What I would probably prefer in these instances - to help with readability, is that either we don't rely on a function here (as there is no parameterization) and that results are just set as globals. If this is only designed to be executed once, then let's just execute it once.

* `validate_private_acl_and_kms_encryption` should be inlined and executed as early as possible.
* The `private` and `encrypted_by_kms` global values should be set in the inlined replacement.
* Rules should be created for both conditions so that tracing will reflect the failure class. You can just add an `is_` prefix for each condition (example: `is_private = rule { private }`).

Reply:
I agree that the function in this case and in some of my new second-generation policies under https://github.com/hashicorp/terraform-guides/tree/master/governance/second-generation are not parameterized and are only invoked once. However, some of those policies do have functions that are parameterized, particularly when they accept a list of allowed values. It is true that in those cases, the functions could just read the global list and that the functions could be inlined to the main policy. Also, I have written other policies in which the functions have more parameters and are invoked more than once.

So, I prefer to do all the processing inside functions so that all the policies follow the same pattern.  I don't see any harm in doing this.

Additionally, in so far as users will copy code from one policy to another, using functions makes it clearer which portion of the source policy should be copied.

Frankly, I find the way Sentinel reports evaluation of rules useless to the consumer of the policies by which I mean the person who ran some Terraform code in a TFE workspace and encountered violations.  One ends up with many lines of useless output and line numbers.  Those might be useful to the writer of the policy, but not to the consumer of its violations.  Using a single main rule instead of multiple rules reduces Sentinel's verbose rule processing output.

Reply:
I think I'd be okay with this if the "result generator" function was not returning an ambiguous list. A map would be better:

```
validate_private_acl_and_kms_encryption = func() {
	result = {
		"private":          true,
		"encrypted_by_kms": true,
	}

	s3_buckets = find_resources_from_plan("aws_s3_bucket")
	for s3_buckets as joined_path, resource_map {
		# Set "private" or "encrypted_by_kms" result map values to
		# false here if there is a violation.
	}

	return result
}

validations = validate_private_acl_and_kms_encryption()
is_private = rule { validations["private"] }
is_encrypted_by_kms = rule { validations["encrypted_by_kms"] }

main = rule {
	is_private and is_encrypted_by_kms
}
```



> Frankly, I find the way Sentinel reports evaluation of rules useless to the consumer of the policies by which I mean the person who ran some Terraform code in a TFE workspace and encountered violations. One ends up with many lines of useless output and line numbers. Those might be useful to the writer of the policy, but not to the consumer of its violations. Using a single main rule instead of multiple rules reduces Sentinel's verbose rule processing output.

We are working on correcting the way trace output is shown within a specific TFE run.

Having **clearly defined rules** is a core design pattern for Sentinel, and our documentation should be promoting these patterns wherever possible, regardless of opinions on the current state of their usefulness.